### ISSUE TITLE
Similarity between lazy bindings and hash codes

### ISSUE THREAD

**Issue by bender2k14** [elmish/Elmish.WPF#117]
There are two lazy bindings: `OneWayLazy` and `OneWaySeqLazy`.  I expect that the following observations hold for both.  For simplicity though, let's just consider `OneWayLazy`.

The simplest way to create a `OneWayLazy` binding is via the method

https://github.com/elmish/Elmish.WPF/blob/5df8514ce4452883a1e417c9f6f479f545339f9e/src/Elmish.WPF/Binding.fs#L245-L261

Intuitively, what would have been a function `get : 'model -> 'b` in the `OneWay` case is decomposed to two functions of types `get : 'model -> 'a` and `map : 'a -> 'b` and joined by an additional function `equals : 'a -> 'a -> bool`.  The idea is that either `map` or the implicit equality comparison for instances of type `'b` (let's call it `equalB`) is expensive to compute but `get` and `equal` are both inexpensive to compute and `(equal a1 a2) = equalB (map a1) (map a2)` holds for all instances `a1` and `a2` of type `'a`.

Given the assumption that `get` or `equalB` is expensive to compute, the standard solution to this problem is to define a function `hash : 'model -> 'hashCode` such that both `hash` and the equality function for instances of type `'hashCode` (let's call it `equalHashCode`) are inexpensive to compute and `not <| equalHashCode (hash m1) (hash m2)` implies `not <| equalB (get m1) (get m2)` for all instances `m1` and `m2` of type `'model`.

I don't have any need for this (right now), or even expect to ever have a need for this.  I just wanted to share this thought.  If someone else wants this functionality, maybe they will see this and comment here.

P.S.
As a thought exercise, it is interesting to consider how to express `OneWayLazy` as this hash code alternative as well as vise versa.  I think each approach have cases for which they are superior to the other alternative.

(This comment was rather technical.  I hope I got all the details correct.)

**Comment by cmeeren** [comment]
When you say `hash : 'model -> 'hashCode`, do you mean `hash : 'a -> 'hashCode`? In other words, what you're saying is that `equals: 'a -> 'a -> bool` can be replaced by `hash: 'a -> 'hashCode`, and then `'hashCode` will be compared using the default equality comparison?

If so, then that's certainly a possibility. Heck, it could even be added as an overload using the existing internals - just define `equals` internally as `fun a1 a2 -> hash a1 = hash a2`.

I don't have any need for it though, nor do I immediately see it simplifying anything since we have the `refEq` and `elmEq` helpers in the library which together cover most use-cases for `equals` in the lazy bindings. `hash` instead of `equals` simply replaces a two-parameter explicit equality function with a single-parameter "`getSomethingToCompare`" function and an implicit equality comparison for the returned value.

**Comment by bender2k14** [comment]
> When you say `hash : 'model -> 'hashCode`, do you mean `hash : 'a -> 'hashCode`?

I don't think so.  Given any `hash2 : 'a -> 'hashCode`, I can create a `hash1 : 'model -> 'hashCode` via `get >> hash 2`, so (of course) this restriction isn't necessary.

We probably have very different examples in our mind for how this would be used.  The standard example I have is...
```
let model = ... // with { Property = { Id, (* lots more * ) } }
let get m = m.Property
let equal a b = a.Id = b.Id
let map = id
```
...which is to say, if computing structural equality is very expensive, just have an ID and check for equality using that instead.  In particular, I am (obviously) not seeing much benefit from `map`.

> In other words, what you're saying is that `equals: 'a -> 'a -> bool` can be replaced by `hash: 'a -> 'hashCode`, and then `'hashCode` will be compared using the default equality comparison?

In this replacement, we also need `equalB : 'b -> 'b -> bool` and need to execute it when the hash codes are equal.

> Heck, it could even be added as an overload using the existing internals - just define equals internally as `fun a1 a2 -> hash a1 = hash a2`.

No, because disequal hash codes imply disequality but equal hash codes imply nothing.

**Comment by cmeeren** [comment]
Sorry, I'm completely lost. Could you give an example using both current and proposed methods?

> if computing structural equality is very expensive, just have an ID and check for equality using that instead.

Just want to add that given a fully immutable model (common in Elmish), reference equality is often a good choice instead of explicit IDs.

**Comment by bender2k14** [comment]
> Sorry, I'm completely lost. Could you give an example using both current and proposed methods?

Sure, don't have time now.  I will get back to this.

> Just want to add that given a fully immutable model (common in Elmish), reference equality is often a good choice instead of explicit IDs.

Indeed.  By that same reasoning, `OneWay` is _typically_ a good choice instead of `OneWayLazy`.  The whole reason for using `OneWayLazy` is that case in question is not _typical_.

**Comment by bender2k14** [comment]
TLDR; I was wrong.

My first mistake was

> ...and `(equal a1 a2) = equalB (map a1) (map a2)` holds for all instances `a1` and `a2` of type `'a`.

We don't want an "[if and only if](https://en.wikipedia.org/wiki/If_and_only_if)", we want an "[implication](https://en.wikipedia.org/wiki/Material_conditional)".  That is, we want that for all instances `a1` and `a2` of type `'a`, it follows that `equal a1 a2` implies `equalB (map a1) (map a2)`.  We want to short circuit computation if  `equal a1 a2` returns `true`.  If it returns `false`, then we can proceed with process of updating the binding.  It might be the case that that the view doesn't change--that we don't raise any change events.  That is ok.

> ...the standard solution to this problem is to define a function `hash : 'model -> 'hashCode` such that both `hash` and the equality function for instances of type `'hashCode` (let's call it `equalHashCode`) are inexpensive to compute and `not <| equalHashCode (hash m1) (hash m2)` implies `not <| equalB (get m1) (get m2)` for all instances `m1` and `m2` of type `'model`.

Since `p => q` (where `=>` here means "implies") is equivalent to `(not q) => (not p)`, an equivalent form of the above implication is for all instances `m1` and `m2` of type `'model`, it follows that `equalB (get m1) (get m2)` implies `equalHashCode (hash m1) (hash m2)`.

Both hash codes and our lazy bindings have one-sided error, but they err on _opposite_ sides.

> > When you say `hash : 'model -> 'hashCode`, do you mean `hash : 'a -> 'hashCode`?
> 
> I don't think so.

Nope.  I was wrong there too.  You were right.

> Sorry, I'm completely lost.

Yea, I can see why now.  Thanks for being so patient with me as I struggled through those thoughts.

The `equals` function in our lazy bindings are allowed to have one-sided err.  If `equals` returns its definite answer (which is that everything is equal), then we halt the update computation.  If it returns its indefinite answer (which is that something _might be_ unequal), then we continue the update process and do a more exhaustive equality checks.

Hashcodes are duals to our lazy bindings.  A typical use case for hashcodes is implementing a structure equality function, such as for a record type.  If a hashcode-based equality function returns its definite answer for some field of a record (which is that two instances of for that field are unequal), then a full structural equality computation on these two instances can be skipped.  If the hashcode-based equality function returns its indefinite answer for some field of a record (which is that two instances _might be_ structurally equal), then the computation must continue and do a more exhaustive equality checks on these instances.  Since record equality is computed by ANDing equality computations for each field, getting a `false` short circuits the computation.

In summary, our lazy bindings are able to short circuit computation when things are equal.   In contrast, hashcodes are used in structural equality functions to short circuit computation when things are unequal.
